aspnetcore/grpc/performance.md

```diff
@@ -0,0 +1,150 @@
+
+::: moniker range=">= aspnetcore-5.0"
+
+.NET 5 introduces the `SocketsHttpHandler.EnableMultipleHttp2Connections` property. When set to true, additional HTTP/2 connections are created by a channel when the concurrent stream limit is reached. When a `GrpcChannel` is created its internal `SocketsHttpHandler` is automatically configured to create additional HTTP/2 connections. If an app configures its own handler, consider setting `EnableMultipleHttp2Connections` to true:
+
+```csharp
+var channel = GrpcChannel.ForAddress("https://localhost", new GrpcChannelOptions
+{
+    HttpHandler = new SocketsHttpHandler
+    {
+        EnableMultipleHttp2Connections = true,
+
+        // ...configure other handler settings
+    }
+});
+```
+
+::: moniker-end
+
+There are a couple of workarounds for .NET Core 3.1 apps:
+
+* Create separate gRPC channels for areas of the app with high load. For example, the `Logger` gRPC service might have very high load. Use a separate channel to create the `LoggerClient` in the app.
+* Use a pool of gRPC channels. A simple way to do this could be to create a list of gRPC channels. `Random` is used to pick a channel from the list each time a gRPC channel is needed. This would randomly distribute calls over multiple connections.
+
+> [!IMPORTANT]
+> Increasing the maximum concurrent stream limit on the server is another way to solve this problem. In Kestrel this is configured with <xref:Microsoft.AspNetCore.Server.Kestrel.Core.Http2Limits.MaxStreamsPerConnection>.
+>
+> Increasing the maximum concurrent stream limit is not recommended. Too many streams on a single HTTP/2 connection introduces new performance issues:
+>
+> * Thread contention between streams trying to write to the connection.
+> * Connection packet loss causes all calls to be blocked at the TCP layer.
+
+::: moniker range=">= aspnetcore-5.0"
+
+## Keep alive pings
+
+Keep alive pings can be used to keep HTTP/2 connections alive during periods of inactivity. Having an existing HTTP/2 connection ready when an app resumes activity allows for the initial gRPC calls to be made quickly, without a delay caused by the connection being reestablished.
+
+Keep alive pings are configured on <xref:System.Net.Http.SocketsHttpHandler>:
+
+```csharp
+var handler = new SocketsHttpHandler
+{
+    PooledConnectionIdleTimeout = Timeout.InfiniteTimeSpan,
+    KeepAlivePingDelay = TimeSpan.FromSeconds(60),
+    KeepAlivePingTimeout = TimeSpan.FromSeconds(30),
+    EnableMultipleHttp2Connections = true
+};
+
+var channel = GrpcChannel.ForAddress("https://localhost:5001", new GrpcChannelOptions
+{
+    HttpHandler = handler
+});
+```
+
+The preceding code configures a channel that sends a keep alive ping to the server every 60 seconds during periods of inactivity. The ping will ensure the server and any proxies in use won't close the connection because of inactivity.
```

Comment:
```suggestion
The preceding code configures a channel that sends a keep alive ping to the server every 60 seconds during periods of inactivity. The ping ensures the server and any proxies in use won't close the connection because of inactivity.
```